Admin/AdminHelper.php

```diff
@@ -108,82 +91,69 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
-
-        try {
-            $value = $fieldDescription->getValue($form->getData());
-        } catch (NoValueException $e) {
-            $value = null;
-        }
-
-        // retrieve the posted data
-        $data = $admin->getRequest()->get($formBuilder->getName());
-
-        if (!isset($data[$childFormBuilder->getName()])) {
-            $data[$childFormBuilder->getName()] = array();
-        }
-
-        $objectCount = count($value);
-        $postCount   = count($data[$childFormBuilder->getName()]);
+        $elementId = preg_replace('#.(\d+)#', '[$1]', implode('.', explode('_', substr($elementId, strpos($elementId, '_') + 1))));
 
-        $fields = array_keys($fieldDescription->getAssociationAdmin()->getFormFieldDescriptions());
-
-        // for now, not sure how to do that
-        $value = array();
-        foreach ($fields as $name) {
-            $value[$name] = '';
-        }
-
-        // add new elements to the subject
-        while ($objectCount < $postCount) {
-            // append a new instance into the object
-            $this->addNewInstance($form->getData(), $fieldDescription);
-            $objectCount++;
-        }
-
-        $this->addNewInstance($form->getData(), $fieldDescription);
+        // append a new instance into the object
+        $this->addNewInstance($admin, $elementId);
 
+        // return new form with empty row
         $finalForm = $admin->getFormBuilder()->getForm();
         $finalForm->setData($subject);
-
-        // bind the data
         $finalForm->setData($form->getData());
 
-        return array($fieldDescription, $finalForm);
+        return $finalForm;
     }
 
     /**
-     * Add a new instance to the related FieldDescriptionInterface value
+     * Add a new instance
      *
-     * @param object                                              $object
-     * @param \Sonata\AdminBundle\Admin\FieldDescriptionInterface $fieldDescription
+     * @param AdminInterface $admin
+     * @param string         $elementId
      *
-     * @throws \RuntimeException
+     * @throws \Exception
      */
-    public function addNewInstance($object, FieldDescriptionInterface $fieldDescription)
+    public function addNewInstance(AdminInterface $admin, $elementId)
     {
-        $instance = $fieldDescription->getAssociationAdmin()->getNewInstance();
-        $mapping  = $fieldDescription->getAssociationMapping();
+        $entity = $admin->getSubject();
 
-        $method = sprintf('add%s', $this->camelize($mapping['fieldName']));
+        $propertyAccessor = new PropertyAccessor();
 
-        if (!method_exists($object, $method)) {
-            $method = rtrim($method, 's');
+        $collection = $propertyAccessor->getValue($entity, $elementId);
 
-            if (!method_exists($object, $method)) {
-                $method = sprintf('add%s', $this->camelize(Inflector::singularize($mapping['fieldName'])));
+        if ($collection instanceof \Doctrine\ORM\ArrayCollection) {
```

Comment:
Just change the whole if block to:

```php

        if ($collection instanceof \Doctrine\ORM\PersistentCollection || $collection instanceof \Doctrine\ODM\MongoDB\PersistentCollection) {
            //since doctrine 2.4
            $entityClassName = $collection->getTypeClass()->getName();
        } elseif ($collection instanceof \Doctrine\Common\Collections\Collection) {
            $entityClassName = $this->entityClassNameFinder($admin, explode('.', preg_replace('#\[\d*?\]#', '', $elementId)));
        } else {
            return;
        }
```

with this change it should work for all possible mappers of doctrine (including the easy check for ORM and ODM)